### ISSUE TITLE
Issue with Automatically Cleared Outputs in Saved Notebooks

### ISSUE THREAD

**Issue by devpkr1** [dbcodeio/public#44]
Dear Mike,

I hope you are doing well. I am writing to report an issue with saved code notebooks that I have recently encountered. Whenever I reopen a saved notebook that contains database-related code, the output cells are automatically cleared. This behavior significantly impacts my ability to review previous outputs, which is crucial for reporting and sharing notebooks with others.

As preserving the outputs is essential for collaboration and accurate reporting, I would appreciate any assistance or guidance on how to prevent this automatic clearing of outputs upon reopening the notebooks.

Thank you for your attention to this matter. I look forward to your support in resolving this issue.

Best regards,
Pradeep Kumar

**Comment by mikeburgh** [comment]
Yeah I had been asked for this previously.. I will look at getting it added, maybe with an option to toggle if you want to store the data or not (security concerns, or risk of people being confused in the case of a "run book" where it needs to be clear if it's already been run). 

Would the ability to password protect it be of any use ?

**Comment by devpkr1** [comment]
Hi Mike,

Thank you for your quick response and for considering the toggle option to store outputs. I think it's a good approach to balance both security concerns and the need for clarity in run books.

As for password protection, having a toggle button or an option that allows users to save outputs either with or without a password would be more appropriate. This way, users can decide based on the specific use case—whether they prioritize security or ease of access when sharing the notebook.

You can add these buttons in the notebook header , I think this could make it more user friendly.

![Screenshot 2024-10-08 235627](https://github.com/user-attachments/assets/a16eede8-780c-4f12-add4-faf666484eed)

I appreciate your attention to this matter and look forward to the implementation.

Best regards,
Pradeep Kumar

**Comment by devpkr1** [comment]
Hi Mike,

I frequently use Jupyter Notebook in VS Code for analysis and machine learning. As you know, when we save a Jupyter Notebook, it retains the output results. This can lead to confusion when someone opens an existing notebook, so it leads the confusion whether the code cells have been executed after opening the notebook, because the output is already present there.

To address this, there’s a tick mark in each cell that only appears when the notebook or a specific cell is executed after opening. If the notebook hasn’t been run, these tick marks are absent, making it easy to distinguish. Similarly, DBCode notebooks also include this tick mark feature, allowing users to clearly identify whether the notebook or a particular cell has been executed. This helps avoid confusion, especially in the case of a “run book,” where it’s essential to know if the notebook has already been run.

I hope this solution helps in resolving the issue and prevents any potential misunderstanding about the state of the notebook.

Thanks and Regards,
Pradeep

**Comment by devpkr1** [comment]
Hi Mike,

I frequently use Jupyter Notebook in VS Code for analysis and machine learning. As you know, when we save a Jupyter Notebook, it retains the output results. This can lead to confusion when someone opens an existing notebook, so it leads the confusion whether the code cells have been executed after opening the notebook, because the output is already present there.

To address this, there’s a tick mark in each cell that only appears when the notebook or a specific cell is executed after opening. If the notebook hasn’t been run, these tick marks are absent, making it easy to distinguish. Similarly, DBCode notebooks also include this tick mark feature, allowing users to clearly identify whether the notebook or a particular cell has been executed. This helps avoid confusion, especially in the case of a “run book,” where it’s essential to know if the notebook has already been run.

I hope this solution helps in resolving the issue and prevents any potential misunderstanding about the state of the notebook.

![Image](https://github.com/user-attachments/assets/95fb7c71-284d-46f5-afa6-6633cc4da050)

Thanks and Regards,
Pradeep

**Comment by mikeburgh** [comment]
Thanks, about to start on some of the notebook stuff so will keep this in mind

**Comment by mikeburgh** [comment]
Version 1.6.8 has been released which has preliminary support for this. State is not restored at the moment, but can be saved. This is required for other NB features I am working on in tandem

**Comment by devpkr1** [comment]
Thank you for the update, Mike! It's great to hear that version 1.6.8 includes preliminary support for saving states. While the state restoration is still pending, it's promising to see this feature being prioritized. Looking forward to seeing how this evolves along with the other NB features you're working on. Appreciate your efforts!

**Comment by mikeburgh** [comment]
No worries.

You can now restore the state. Your notebook might have an issue restoring if it was saved with output from before as the structure had to change a little.

Let me know if you find anything, otherwise I will close this in a a few days.

**Comment by devpkr1** [comment]
It’s working perfectly, and I’ve tested it thoroughly. Not only does it restore the output, but it also restores charts, which is fantastic.

However, when opening a saved notebook, tick marks appear in the code cells with outputs, even if no cells have been executed after opening. I understand this happens because the code cells are automatically processed to retrieve the outputs.

This can create confusion, as it gives the impression that the cells have been recently executed. Ideally, tick marks should only appear when the cells are explicitly executed after opening the notebook to avoid this ambiguity.

Nonetheless, the current setup works perfectly. Thank you for your continuous efforts to make database management simple and straightforward.

Thanks,
Pradeep

**Comment by mikeburgh** [comment]
Sorry I miss understood your previous comment, I thought you wanted the execution summary saved for reloading, agree it makes more sense not to be saved/restored, so I have changed it to that. 

I will close this when it's released next week.

**Comment by mikeburgh** [comment]
Released in 1.6.10
